crates/bevy_reflect/src/type_data.rs

```diff
@@ -0,0 +1,63 @@
+use downcast_rs::{impl_downcast, Downcast};
+
+/// A trait for types generated by the [`#[reflect_trait]`][0] attribute macro.
+///
+/// [0]: crate::reflect_trait
+pub trait TypeData: Downcast + Send + Sync {
+    fn clone_type_data(&self) -> Box<dyn TypeData>;
+}
+impl_downcast!(TypeData);
+
+impl<T: 'static + Send + Sync> TypeData for T
+where
+    T: Clone,
+{
+    fn clone_type_data(&self) -> Box<dyn TypeData> {
+        Box::new(self.clone())
+    }
+}
+
+/// Trait used to generate [`TypeData`] for trait reflection.
+///
+/// This is used by the `#[derive(Reflect)]` macro to generate an implementation
+/// of [`TypeData`] to pass to [`TypeRegistration::insert`].
+pub trait FromType<T> {
+    fn from_type() -> Self;
+}
+
+pub trait SpecializedFromType<T>: FromType<T> {
```

Comment:
Could you add docs explaining what this struct is and how to use it? Including documentation on the method as well.

Reply:
Done.